test/component/storage/storage.go

```diff
@@ -764,27 +781,69 @@ func TestStoreGetBulk(t *testing.T, provider spi.Provider) { //nolint: funlen //
 		require.Equal(t, "value1", string(values[0]))
 		require.Equal(t, "value2", string(values[1]))
 	})
-	t.Run("One value found, one not", func(t *testing.T) {
-		store, err := provider.OpenStore(randomStoreName())
-		require.NoError(t, err)
-		require.NotNil(t, store)
+	t.Run("Two values found, one not", func(t *testing.T) {
+		t.Run("Value not found was the second one", func(t *testing.T) {
+			store, err := provider.OpenStore(randomStoreName())
+			require.NoError(t, err)
+			require.NotNil(t, store)
 
-		defer func() {
-			require.NoError(t, store.Close())
-		}()
+			defer func() {
+				require.NoError(t, store.Close())
+			}()
 
-		err = store.Put("key1", []byte("value1"),
-			[]spi.Tag{
-				{Name: "tagName1", Value: "tagValue1"},
-				{Name: "tagName2", Value: "tagValue2"},
-			}...)
-		require.NoError(t, err)
+			err = store.Put("key1", []byte("value1"),
+				[]spi.Tag{
+					{Name: "tagName1", Value: "tagValue1"},
+					{Name: "tagName2", Value: "tagValue2"},
+				}...)
+			require.NoError(t, err)
 
-		values, err := store.GetBulk("key1", "key2")
-		require.NoError(t, err)
-		require.Len(t, values, 2)
-		require.Equal(t, "value1", string(values[0]))
-		require.Nil(t, values[1])
+			err = store.Put("key2", []byte("value2"),
+				[]spi.Tag{
+					{Name: "tagName1", Value: "tagValue1"},
+					{Name: "tagName2", Value: "tagValue2"},
+				}...)
+			require.NoError(t, err)
+
+			values, err := store.GetBulk("key1", "nonexistentkey", "key2")
+			require.NoError(t, err)
+
+			require.Len(t, values, 3)
+			require.Equal(t, "value1", string(values[0]))
+			require.Nil(t, values[1])
+			require.Equal(t, "value2", string(values[2]))
+		})
+		t.Run("Value not found was the third one", func(t *testing.T) {
+			store, err := provider.OpenStore(randomStoreName())
```

Comment:
There's a fair bit of duplication between these tests, when it seems like what you want to do is go through a test matrix and verify all combinations of found / not found

Reply:
I agree - I have a bunch of other tests that could benefit from using a test table/matrix too. I created a follow-up issue since refactoring all the tests is a giant task: #2930